pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -1781,22 +1781,12 @@ func (configgen *ConfigGeneratorImpl) onVirtualOutboundListener(
 		Protocol: protocol.TCP,
 	}
 
-	if len(ipTablesListener.FilterChains) < 1 || len(ipTablesListener.FilterChains[0].Filters) < 1 {
-		return ipTablesListener
-	}
-
-	// contains all filter chains except for the final passthrough/blackhole
-	initialFilterChain := ipTablesListener.FilterChains[:len(ipTablesListener.FilterChains)-1]
-
-	// contains just the final passthrough/blackhole
-	fallbackFilter := ipTablesListener.FilterChains[len(ipTablesListener.FilterChains)-1].Filters[0]
-
 	if util.IsAllowAnyOutbound(node) {
 		svc = util.FallThroughFilterChainPassthroughService
 	}
 
 	pluginParams := &plugin.InputParams{
-		ListenerProtocol:           plugin.ListenerProtocolTCP,
+		ListenerProtocol:           plugin.ListenerProtocolAuto,
```

Comment:
This is even harder to maintain. IIUC you're trying to set `ListenerProtocol` to `plugin.ListenerProtocolAuto` so that you can force `plugin.OnVirtualListener` to call `plugin.OnOutboundListener` with protocol set to `ListenerProtocolAuto` and then undoing that by reseting protocol back to `ListenerProtocolTCP`.

Why not move the logic from https://github.com/istio/istio/blob/1679610e0279a7285eba6beb4c80bce1e9292ea6/pilot/pkg/networking/plugin/mixer/mixer.go#L169-L179 into a common utility and call it from both `OnOutboundListener` & `OnVirtualListener`

Reply:
I am setting per filter chain protocol. This is how the http sniffing logic works. It sets the listener protocol to auto, and sets the per filter chain protocol to tcp/http. If it was upto me, I would completely nuke the onVirtualListener and simply use the onOutboundListener as both are dealing with outbound (onvirtualListener is badly named as it stands today because we have an inbound and outbound virtual listener).

Reply:
I see, when I initially added this code, based on your recommendation we augment the plugin interface. After the virtual outbound and inbound were split, the interface was not augmented.

I'm still little confused if this actually works. I don't think you will end up with an extra mixer filter here as `mutable.FilterChains[cnum].ListenerProtocol` is not set for the passed in `fallthrough` filter chain. If you can paste the Envoy config dump for `0.0.0.0_15001` it would be very helpful.

Reply:
I see, when I initially added this code, based on your recommendation we augment the plugin interface. After the virtual outbound and inbound were split, the interface was not augmented.

I'm still little confused if this actually works. I don't think you will end up with an extra mixer filter here as `mutable.FilterChains[cnum].ListenerProtocol` is not set for the passed in `fallthrough` filter chain. If you can paste the Envoy config dump for `0.0.0.0_15001` it would be very helpful.